### ISSUE TITLE
llm_local: reuse the KV cache across ReAct iterations instead of re-prefilling the whole transcript

### ISSUE THREAD

**Issue by fpt** [fpt/rs-gallium#86]
## What happens now

`llm_local::generate` builds a fresh `LlamaContext` for every model call
(`crates/gallium-agent/src/llm_local.rs:560`, and again at `:891` for the media
path) and drops it when the call returns. Every ReAct iteration therefore
re-evaluates the entire transcript from token 0, even though iteration *N*'s
prompt is a prefix of iteration *N+1*'s.

## What it costs

Measured on `gemma-4-E4B-it-Q4_K_M`, Metal, M3 — a two-iteration turn with one
`Bash` call, using the timing added in #85:

```
⏱  ttft 95.45s · prefill 11843 tok 124.1 tok/s · decode 11 tok 13.8 tok/s
⏱  ttft 97.75s · prefill 11882 tok 121.6 tok/s · decode 30 tok 10.0 tok/s
📊 tokens: in=23725, out=43, total=23768 · ttft 95.45s · prefill 122.8 tok/s · decode 10.8 tok/s
```

**193 seconds of prefill against 3 seconds of generation.** The second call
re-processed the same ~11.8k tokens the first one had just processed, to append
about 40. The prompt is already large before the user says anything interesting
— system prompt, `CLAUDE.md`, skill catalog, tool schemas — so this is the
normal case for an agent turn, not a pathological one.

On this turn, reuse should take the second iteration's prefill from ~97s to the
cost of its suffix, i.e. roughly a 2× turn-level win, and more on longer turns:
the waste grows with iteration count *and* with transcript length. The same
applies **across turns** in a REPL session, not only within one.

## Sketch

Hold one `LlamaContext` per conversation, and on each call:

1. tokenize the new prompt;
2. find the longest common token prefix with what the context already holds;
3. `clear_kv_cache_seq(seq, common_len, -1)` to drop the divergent tail;
4. decode only the suffix.

`llama-cpp-2` 0.1.151 exposes what this needs — `clear_kv_cache_seq`,
`kv_cache_seq_add`, `copy_kv_cache_seq`, `kv_cache_seq_pos_max`.

A **token-level** prefix comparison rather than an assumed append is the point.
The next prompt is re-rendered through the jinja template from scratch, and the
assistant turn inside it is gallium's *re-serialization* of the tool call
(`serialize_tool_calls`), not the tokens the model actually emitted. Those can
differ. A cache reused past the real divergence point produces plausible wrong
logits, which is invisible from outside.

## What this touches

- **Concurrency.** `LlamaLocalProvider` is `Send + Sync` today on the explicit
  grounds that nothing is mutated per call — the comment above
  `unsafe impl Sync` (`:108`) says so, and the app-server runs concurrent turns
  against one provider. A retained context is per-conversation mutable state, so
  it needs a mutex, or one context per thread, and that decision should be made
  deliberately rather than fall out of the change.
- **`n_ctx` sizing.** The context is currently sized per call as
  `n_ctx.max(n_prompt + max_tokens)` from a hardcoded `LOCAL_CONTEXT_WINDOW`
  (8192) — `[llm] contextWindow` never reaches llama.cpp. A retained context has
  to be sized once, up front, which means that knob has to actually arrive, plus
  a defined behaviour when the transcript outgrows it (rebuild and re-prefill is
  a fine first answer, as long as it is not silent).
- **Compaction.** `runtime::run_turn` drops messages from the *front* of the
  history between turns, which invalidates the cached prefix wholesale. Reuse
  and compaction need to agree on what happens then.
- **The mtmd path.** `eval_chunks` interleaves projector output with text
  decodes. Text-only reuse first is a reasonable scope cut.
- **The candle backend has the same shape.** `gallium_core::generate` opens with
  `model.reset()` (`crates/gallium-core/src/model.rs:39`) and prefills the whole
  prompt. Same win available; separate change.

## Testing

The sampler seed is fixed at 1234, so with a fixed prompt and temperature a run
is deterministic. That gives a strong and cheap correctness test: a turn with
reuse enabled must produce a **token-identical** stream to the same turn with it
disabled. A prefix bug shows up as a diff rather than as a subtly worse answer.

## Why now

It reshapes the surface that [docs/OPTIMIZATION.md](../blob/main/docs/OPTIMIZATION.md)
proposes to search. Right now prefill is ~97% of an agent turn, so a parameter
search would be tuning almost entirely for prompt evaluation; after reuse, the
balance shifts and the optimum very likely moves with it. Better to land this
before a large study than to redo one — see that document's open questions.
